src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs

```diff
@@ -570,62 +570,62 @@ private SwitchExpressionSyntax ParseSwitchExpression(ExpressionSyntax governingE
-                if (switchExpressionCase.Width == 0 && this.CurrentToken.Kind != SyntaxKind.CommaToken)
-                    break;
-
-                arms.Add(switchExpressionCase);
-                if (this.CurrentToken.Kind != SyntaxKind.CloseBraceToken)
+                var arms = _pool.AllocateSeparated<SwitchExpressionArmSyntax>();
+
+                while (this.CurrentToken.Kind != SyntaxKind.CloseBraceToken)
                 {
-                    var commaToken = this.CurrentToken.Kind == SyntaxKind.SemicolonToken
-                        ? this.EatTokenAsKind(SyntaxKind.CommaToken)
-                        : this.EatToken(SyntaxKind.CommaToken);
-                    arms.AddSeparator(commaToken);
+                    // Help out in the case where a user is converting a switch statement to a switch expression. Note:
+                    // `default(...)` and `default` will also be consumed as a legal syntactic patterns (though the
+                    // latter will fail during binding).  So if the user has `default:` we will recover fine as we
+                    // handle the errant colon below.
+                    var errantCase = this.CurrentToken.Kind == SyntaxKind.CaseKeyword
+                        ? AddError(this.EatToken(), ErrorCode.ERR_BadCaseInSwitchArm)
+                        : null;
+
+                    var savedState = _termState;
+                    _termState |= TerminatorState.IsPatternInSwitchExpressionArm;
+                    var pattern = ParsePattern(Precedence.Coalescing, whenIsKeyword: true);
+                    _termState = savedState;
+
+                    // We use a precedence that excludes lambdas, assignments, and a conditional which could have a
+                    // lambda on the right, because we need the parser to leave the EqualsGreaterThanToken to be
+                    // consumed by the switch arm. The strange side-effect of that is that the conditional expression is
+                    // not permitted as a constant expression here; it would have to be parenthesized.
+
+                    var switchExpressionCase = _syntaxFactory.SwitchExpressionArm(
+                        pattern,
+                        ParseWhenClause(Precedence.Coalescing),
+                        // Help out in the case where a user is converting a switch statement to a switch expression.
+                        // Consume the `:` as a `=>` and report an error.
+                        this.CurrentToken.Kind == SyntaxKind.ColonToken
+                            ? this.EatTokenAsKind(SyntaxKind.EqualsGreaterThanToken)
+                            : this.EatToken(SyntaxKind.EqualsGreaterThanToken),
+                        ParseExpressionCore());
+
+                    // If we're not making progress, abort
+                    if (errantCase is null && switchExpressionCase.FullWidth == 0 && this.CurrentToken.Kind != SyntaxKind.CommaToken)
+                        break;
```

Comment:
the bug was here.  it was possible to consume the errant-case but then discard it, leading to it being dropped from the final tree.